core/src/main/java/jenkins/install/UpgradeWizard.java

```diff
@@ -0,0 +1,146 @@
+package jenkins.install;
+
+import hudson.Extension;
+import hudson.model.PageDecorator;
+import hudson.model.UpdateSite.Plugin;
+import hudson.util.HttpResponses;
+import hudson.util.VersionNumber;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.*;
+import static org.apache.commons.io.FileUtils.*;
+import static org.apache.commons.lang.StringUtils.*;
+
+/**
+ * This is a stop-gap measure until JENKINS-33663 comes in.
+ * This call may go away. Please don't use it from outside.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+@Restricted(NoExternalUse.class)
+@Extension
+public class UpgradeWizard extends PageDecorator {
+    @Inject
+    private Jenkins jenkins;
+
+    /**
+     * Is this instance fully upgraded?
+     */
+    private volatile boolean upToDate;
+
+    /**
+     * File that captures the state of upgrade.
+     *
+     * This file records the vesrion number that the installation has upgraded to.
+     */
+    /*package*/ File getStateFile() {
+        return new File(Jenkins.getInstance().getRootDir(),"upgraded");
+    }
+
+    public UpgradeWizard() throws IOException {
+        updateUpToDate();
+    }
+
+    private void updateUpToDate() throws IOException {
+        upToDate = new VersionNumber("2.0").compareTo(getCurrentLevel()) <= 0;
+    }
+
+    /**
+     * What is the version the upgrade wizard has run the last time and upgraded to?
+     */
+    private VersionNumber getCurrentLevel() throws IOException {
+        VersionNumber from = new VersionNumber("1.0");
+        File state = getStateFile();
+        if (state.exists()) {
+            from = new VersionNumber(defaultIfBlank(readFileToString(state), "1.0").trim());
+        }
+        return from;
+    }
```

Comment:
Any file corruption or accessibility issue moves the user back to "1.0" and displays the notice bar.
I'm not sure this approach is correct. Maybe a virtual "2.0" plugin is a better solution.